internal/pipeline/pipeline_test.go

```diff
@@ -315,3 +317,266 @@ func isPipelineEqual(t *testing.T, p1, p2 Pipeline) bool {
+
+	for _, n := range inputs {
+		err := ValidateFunctionName(n.Name)
+		if n.Valid && err != nil {
+			t.Fatalf("function name %s should be valid", n.Name)
+		}
+		if !n.Valid && err == nil {
+			t.Fatalf("function name %s should not be valid", n.Name)
+		}
+	}
+}
+
+func TestPipelineValidate(t *testing.T) {
+	type input struct {
+		Name  string
+		Input string
+		Valid bool
+	}
+	cases := []input{
+		{
+			Name: "no sources, no functions",
+			Input: `
+apiVersion: kpt.dev/v1alpha1
+kind: Pipeline
+metadata:
+  name: pipeline
+`,
+			Valid: true,
+		},
+		{
+			Name: "have sources, no functions",
+			Input: `
+apiVersion: kpt.dev/v1alpha1
+kind: Pipeline
+metadata:
+  name: pipeline
+sources:
+- ./base
+`,
+			Valid: true,
+		},
+		{
+			Name: "have sources and functions",
+			Input: `
+apiVersion: kpt.dev/v1alpha1
+kind: Pipeline
+metadata:
+  name: pipeline
+sources:
+- ./base
+- ./*
+
+generators:
+- image: gcr.io/kpt-functions/generate-folders
+  config:
+    apiVersion: cft.dev/v1alpha1
+    kind: ResourceHierarchy
+    metadata:
+      name: root-hierarchy
+      namespace: hierarchy # {"$kpt-set":"namespace"}
+transformers:
+- image: patch-strategic-merge
+  configPath: ./patch.yaml
+- image: gcr.io/kpt-functions/set-annotation
+  configMap:
+    environment: dev
+
+validators:
+- image: gcr.io/kpt-functions/policy-controller-validate
+`,
+			Valid: true,
+		},
+		{
+			Name: "invalid apiversion",
+			Input: `
+apiVersion: kpt.dev/v1
+kind: Pipeline
+metadata:
+  name: pipeline
+sources:
+- ./base
+`,
+			Valid: false,
+		},
+		{
+			Name: "absolute source path",
+			Input: `
+apiVersion: kpt.dev/v1alpha1
+kind: Pipeline
+metadata:
+  name: pipeline
+sources:
+- /foo/bar
+`,
+			Valid: false,
+		},
+		{
+			Name: "invalid function name",
+			Input: `
+apiVersion: kpt.dev/v1alpha1
+kind: Pipeline
+metadata:
+  name: pipeline
+sources:
+- ./*
+transformers:
+- image: patch@_@strategic-merge
+  configPath: ./patch.yaml
+`,
+			Valid: false,
+		},
+		{
+			Name: "more than 1 config",
+			Input: `
+apiVersion: kpt.dev/v1alpha1
+kind: Pipeline
+metadata:
+  name: pipeline
+sources:
+- ./*
+transformers:
+- image: patch-strategic-merge
+  configPath: ./patch.yaml
+  configMap:
+    environment: dev
+`,
+			Valid: false,
+		},
+		{
+			Name: "absolute config path",
+			Input: `
+apiVersion: kpt.dev/v1alpha1
+kind: Pipeline
+metadata:
+  name: pipeline
+sources:
+- ./*
+transformers:
+- image: patch-strategic-merge
+  configPath: /patch.yaml
+`,
+			Valid: false,
+		},
+	}
+
+	for _, c := range cases {
+		b := bytes.NewBufferString(c.Input)
+		// FromReader will validate the pipeline
+		_, err := FromReader(b)
+		if c.Valid && err != nil {
+			t.Fatalf("%s: pipeline should be valid, %s", c.Name, err)
+		}
+		if !c.Valid && err == nil {
+			t.Fatalf("%s: pipeline should not be valid", c.Name)
+		}
+	}
+}
+
+func TestValidatePath(t *testing.T) {
+	type input struct {
+		Path  string
+		Valid bool
+	}
+
+	cases := []input{
+		{
+			Path:  "a/b/c",
```

Comment:
Currently this allows having subdirectory named `*`.  Is that intentional?

Can you think of other corner cases?

Reply:
`*` is a valid file name in Unix/Linux. It's possible that there is a subdirectory `./a/*` so I think it's valid.

Reply:
A more reasonable scenario is the intended `./*` but didn't get it quite right (e.g. `*`, or `./*`), as opposed to having a directory with an asterisk which is extremely unlikely.

Reply:
I think this is about how restrict this validation function should be. Is a path to a non-exist directory valid? Is a path to a invalid kpt package valid? Is the accidental typo in path but that path is a kpt package valid? I think these scenarios are out of this functions's duty. If there are errors, they will be reported when kpt try to read and process the sources and we shouldn't check every possible situation. If we find that some scenarios happen frequently we can add more restrictions in the future.

Reply:
IMO this layer should be responsible for lexical validation. Concerns like does package actually exists is a separate concern. The comments here are only concerned with lexical correctness.